### PR TITLE
[BEAM-7009] Add Go SDK Standard Coders yaml tests.

### PR DESCRIPTION
**Pull request by lostluck**
This PR has the Go SDK join Java and Python in having its coders tested against the standard_coder.yaml suite.

* It adds representation markers for standard types that previously didn't have representation (such as iterables, rows, length prefixes, etc) in the graph/coder package.
* It adds execution wrappers for encoding, decoding those values in the exec package.
* It adds serialization support for these representations in the graphx package.
* Adds hashing support for row coded values.

Testing of the code here is largely through the newly introduced fromyaml test harness which reads standard_coder.yaml and helps it along a touch, in particular in how to interpret the yaml versions of expected structures, which don't map to go types exactly.

While this PR plumbs in Timers and ParamWindowCoders, these coders are not implemented in the SDK at this time, and those tests are skipped.  At best, the implementation is vestigial, but leaving them wholely unimplemented caused certain difficulties with the standard yaml test.
Similarly, while Windows are now fully plumbed through the stack, this is largely for convenience when testing those codings vs the standard_coders.yaml file. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
This branch was synced to the bad branch from earlier, hence the check failures. I'll rebase with not other changes on monday.

**Comment by lostluck** [comment]
R: @youngoli  

I see the new test failed, I don't anticipate any major changes while I investigate and fix that. PTAL.

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Python PreCommit

**Comment by lostluck** [comment]
R: @tysonjh @youngoli 
cc: @TheNeuralBit 

I'm finally getting around to fixing the build dep issue with this one, but that will not affect it's ability to be reviewed.

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by youngoli** [comment]
Progress update: Started reviewing this today, about 2/3 of the way through. Wanted to let you know I was working on it, so you don't just get radio silence.

**Comment by lostluck** [comment]
Thanks! I appreciate that.

File: sdks/go/test/regression/coders/fromyaml/fromyaml.go
Comment (youngoli) [inline sdks/go/test/regression/coders/fromyaml/fromyaml.go:338]: Is this assumption made specifically based on the yaml file this is meant to be used with? Or is this based on something in the schema spec I'm forgetting?

**Review by youngoli — COMMENTED** [review]
Done reviewing. Most of my comments are either questions or accidental debug printfs, but overall I didn't see anything big.

Also it looks like some python/java tests are failing. I'm guessing they're unrelated?

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go:93]: "Custom schema coding uses the beam row coder, but wrapping it in a length prefix and treats it as an opaque coder."

Does that make more sense?

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go:681]: In this case it's to set it to the 0th window element. In this case it's a small hack to support "window's as element values" which is only necessary to dramatically simplify the yaml tests which actually test the coder values here. 

Deserves a comment to explain it, as it is a hack.

File: sdks/go/test/regression/coders/fromyaml/fromyaml.go
Comment (lostluck) [reply sdks/go/test/regression/coders/fromyaml/fromyaml.go:338]: The yaml file exclusively uses lists of strings, which is why we know what type to interpret them as here.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go:93]: Yeah that makes more sense to me.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go:681]: Ah that makes sense now. Yeah definitely could use a comment.

File: sdks/go/test/regression/coders/fromyaml/fromyaml.go
Comment (youngoli) [reply sdks/go/test/regression/coders/fromyaml/fromyaml.go:338]: Gotcha. It seems a little brittle (if the yaml file ever adds an example with a list of other types then this will break), so a comment mentioning that it's due to the contents of the yaml file would be helpful if that happens.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go:203]: That's right the encoding format for both is the same. I'll put the comment on the array decoder itself rather than here though.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go]: Done.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go]: Done.

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go]: Ah no difference TBH. DecodeVarInt basically just does an extra cast from uint64 to int64, which I didn't remove. 
I went back to `DecodeVarUint64` since it avoids an extra layer of if checks and a function call.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done.

**Comment by lostluck** [comment]
PTAL

**Review by youngoli — APPROVED** [review]
Looks good to me.

**Comment by lostluck** [comment]
It shouldn't be hard to fix, but apparently a new test case was added to standard_coders.yaml was added in the month this PR has sat around, causing it to break. I'll resolve and if it requires a significant change to this PR I'll let you know. Thanks!

**Comment by lostluck** [comment]
No major changes. I've hacked in to ignore tests with logical type for now to get this in, as that's something I want to handle in a different PR. Will merge after the test passes.

**Comment by lostluck** [comment]
All the jenkins tests are passing, and this doesn't change any python code, so the failing python checks are likely spurious.

**Comment by TheNeuralBit** [comment]
Sorry about that @lostluck, I should've let you know about the new test case. There should probably be an easier way for SDKs to temporarily opt out of particular tests. WDYT about giving each suite labels that SDKs can use in an ignorelist?

**Comment by lostluck** [comment]
No worries. I assumed a conflict would happen somewhere. I'm just glad it
was caught before merging.
Supported/ignored labels/urns would be the way to go, but It does mean
redundancy in the test cases making them a touch harder to add/maintain.
This one was caught out since it's about a feature inside the schema coder
spec that can't be detected by the base URN.

Perhaps a new optional field? Most of them can be caught by the URNs
themselves without special handling but it's the nested feature support
that's trickier to work around (like Go looking up "logical" in the
payload).

**Pull request merged**
